mitmproxy/addons/proxyserver.py

```diff
@@ -400,23 +269,28 @@ def inject_tcp(self, flow: Flow, to_client: bool, message: bytes):
         except ValueError as e:
             ctx.log.warn(str(e))
 
-    def server_connect(self, ctx: server_hooks.ServerConnectionHookData):
-        assert ctx.server.address
-        # FIXME: Move this to individual proxy modes.
-        self_connect = ctx.server.address[1] in (
-            self.options.dns_listen_port,
-            self.options.listen_port,
-        ) and ctx.server.address[0] in (
-            "localhost",
-            "127.0.0.1",
-            "::1",
-            self.options.listen_host,
-            self.options.dns_listen_host,
-        )
-        if self_connect:
-            ctx.server.error = (
-                "Request destination unknown. "
-                "Unable to figure out where this request should be forwarded to."
-            )
-        if ctx.server.sockname is None:
-            ctx.server.sockname = self.connect_addr
+    def server_connect(self, data: server_hooks.ServerConnectionHookData):
+        if data.server.sockname is None:
+            data.server.sockname = self._connect_addr
+
+        # Prevent mitmproxy from recursively connecting to itself.
+        assert data.server.address
+        connect_host, connect_port, *_ = data.server.address
+
+        for server in self.servers.values():
+            for listen_host, listen_port, *_ in server.listen_addrs:
```

Comment:
Ideally the server instance should also have a `transport_protocol` property that get's checked for equality. It could very well be the case, that on the other transport someone else is listening, e.g. mitmproxy QUIC on UDP 443, but apache HTTP on TCP 443).